mautic/mautic
# Froala feature flag - disabled by default

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [security precaution]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [y]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes https://mautic.atlassian.net/browse/TPROD-441



#### Description:

The [Froala Javascript library](https://froala.com/) Mautic [use](https://github.com/mautic/mautic/tree/5.x/app/bundles/CoreBundle/Assets/js/libraries/froala) has several security vulnerabilities reported:

https://security.snyk.io/package/npm/froala-editor/2.4.2

It's used in the legacy builder but it seems that many community members are still using it:

https://forum.mautic.org/t/is-anyone-still-using-the-legacy-email-and-page-builder/27819

So the solution in between that will disable this vulnerable library for Mautic 5 but allows users to enable it if they need it and are fine with the security risks is to add a new feature flag (= configuration option). This feature flag is OFF by default so all Mautic administrators must enable it and agree with the downsides.

![Screenshot 2023-05-31 at 13 56 17](https://github.com/mautic/mautic/assets/1235442/70ca2bbc-a60b-489f-81e2-ed482c00dfe2)

If the GrapesJS builder plugin is disabled and a Mautic user tries to open the legacy builder, they get this warning with instructions about what to do:

![Screenshot 2023-05-31 at 13 55 52](https://github.com/mautic/mautic/assets/1235442/c9442e2f-8513-43ac-a199-ddebd846f6b0)

Apart from making Mautic 5 a bit safer, it will also help with performance.

With Froala enabled:
![Screenshot 2023-05-31 at 14 45 29](https://github.com/mautic/mautic/assets/1235442/88f635ea-3b56-438e-9ffe-4abe1aa60ff3)

With Froala disabled:
![Screenshot 2023-05-31 at 14 45 43](https://github.com/mautic/mautic/assets/1235442/711714af-b4dc-4267-b123-3f73a25a62ac)

This feature flag will save 1.6 MB off of the Mautic administration page load. Saved 27% of loaded size and the loading is 11% faster than before.

_The savings will be greater when we remove the Froala library completely. The CSS/LESS files are baked into the library.css file and cannot be easily extracted. So those are still loading all the time._

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Disable the GrapesJS plugin so the email and page builders would use the legacy builder.
3. Open JS console in your browser dev tools to see if there are any JS errors.
4. Try to open the legacy builder. <<< You should get the warning.
5. Go to Global configuration, System settings and enable the Froala library.
6. Open the legacy builder now. <<< The builder works correctly.
